cmd/gcp/secret_manager.go

```diff
@@ -0,0 +1,141 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"hash/crc32"
+	"io"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
+)
+
+// Signer implements crypto.Signer using Google Cloud Secret Manager.
+type Signer struct {
+	publicKey  crypto.PublicKey
+	privateKey crypto.PrivateKey
+}
+
+// Public returns the public key stored in the Signer object.
+func (s *Signer) Public() crypto.PublicKey {
+	return s.publicKey
+}
+
+// Sign signs digest with the private key stored in Google Cloud Secret Manager.
+// Only crypto.SHA256 and ECDSA are supported.
+func (s *Signer) Sign(rand io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+	// Verify hash function and digest bytes length.
+	if opts == nil || opts.HashFunc() != crypto.SHA256 {
+		return nil, fmt.Errorf("unsupported hash func: %v", opts.HashFunc())
+	}
+	if len(digest) != opts.HashFunc().Size() {
+		return nil, fmt.Errorf("digest bytes length %d does not match hash function bytes length %d", len(digest), opts.HashFunc().Size())
+	}
+
+	privateKey, ok := s.privateKey.(*ecdsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("the key stored in Secret Manager is not an ECDSA key")
+	}
+
+	return ecdsa.SignASN1(rand, privateKey, digest)
+}
+
+// NewSecretManagerSigner creates a new signer that uses the ECDSA P-256 key pair in
+// Google Cloud Secret Manager for signing digests.
+func NewSecretManagerSigner(ctx context.Context, publicKeySecretName, privateKeySecretName string) (*Signer, error) {
+	client, err := secretmanager.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create secret manager client: %w", err)
+	}
+	defer client.Close()
+
+	// Public Key
+	publicKeyRaw, err := accessSecretVersion(ctx, client, publicKeySecretName)
+	if err != nil {
+		return nil, err
+	}
+	pemBlock, rest := pem.Decode([]byte(publicKeyRaw))
+	if pemBlock == nil {
+		return nil, errors.New("failed to decode PEM")
+	}
+	if len(rest) > 0 {
+		return nil, fmt.Errorf("extra data after decoding PEM: %v", rest)
+	}
+	var publicKey crypto.PublicKey
+	switch pemBlock.Type {
+	case "PUBLIC KEY":
+		publicKey, err = x509.ParsePKIXPublicKey(pemBlock.Bytes)
+	default:
+		return nil, fmt.Errorf("unsupported PEM type: %s", pemBlock.Type)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	// Private Key
+	privateKeyRaw, err := accessSecretVersion(ctx, client, privateKeySecretName)
+	if err != nil {
+		return nil, err
+	}
+	pemBlock, rest = pem.Decode([]byte(privateKeyRaw))
+	if pemBlock == nil {
+		return nil, errors.New("failed to decode PEM")
+	}
+	if len(rest) > 0 {
+		return nil, fmt.Errorf("extra data after decoding PEM: %v", rest)
+	}
```

Comment:
You could probably factor this PEM decoding bit out into `accessSecretVersion` as it's common between here and the public key version - maybe the method could be `retrieveSecretPEM` or similar?

Reply:
Done.